datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssigner.java

```diff
@@ -5,30 +5,104 @@
  */
 package com.linkedin.datastream.server.assignment;
 
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
-
-import org.apache.commons.lang.NotImplementedException;
+import java.util.stream.Collectors;
 
 import com.linkedin.datastream.server.ClusterThroughputInfo;
+import com.linkedin.datastream.server.DatastreamGroupPartitionsMetadata;
 import com.linkedin.datastream.server.DatastreamTask;
+import com.linkedin.datastream.server.DatastreamTaskImpl;
+import com.linkedin.datastream.server.PartitionThroughputInfo;
 
 
 /**
  * Performs partition assignment based on partition throughput information
  */
 public class LoadBasedPartitionAssigner {
+
+  private static final Integer DEFAULT_KB_RATE = 5;
+  private static final Integer DEFAULT_MESSAGE_RATE = 5;
+
   /**
-   * Get partition assignment
-   * @param throughputInfo Per-partition throughput information
-   * @param assignedPartitions List of assigned partitions
-   * @param unassignedPartitions List of unassigned partitions
-   * @param taskCount Task count
-   * @return Partition assignment
+   * Performs partition assignment based on partition throughput information
+   * @param throughputInfo Per partition throughput information
+   * @param currentAssignment Current assignment
+   * @param unassignedPartitions Unassigned partitions
+   * @param partitionMetadata Partition metadata
+   * @return New assignment
    */
   public Map<String, Set<DatastreamTask>> assignPartitions(ClusterThroughputInfo throughputInfo,
-      List<String> assignedPartitions, List<String> unassignedPartitions, int taskCount) {
-    throw new NotImplementedException();
+      Map<String, Set<DatastreamTask>> currentAssignment, List<String> unassignedPartitions,
+      DatastreamGroupPartitionsMetadata partitionMetadata) {
+    String datastreamGroupName = partitionMetadata.getDatastreamGroup().getName();
+    Map<String, PartitionThroughputInfo> partitionInfoMap = throughputInfo.getPartitionInfoMap();
+
+    // filter out all the tasks for the current datastream group, and retain assignments in a map
+    Map<String, Set<String>> newPartitions = new HashMap<>();
+    currentAssignment.values().forEach(tasks -> tasks.forEach(task -> {
+      if (task.getTaskPrefix().equals(datastreamGroupName)) {
+        Set<String> retainedPartitions = new HashSet<>(task.getPartitionsV2());
+        retainedPartitions.retainAll(partitionMetadata.getPartitions());
+        newPartitions.put(task.getId(), retainedPartitions);
+      }
+    }));
+
+    // sort the current tasks on total throughput
+    Map<String, Integer> taskThroughputMap = new HashMap<>();
+    PartitionThroughputInfo defaultPartitionInfo = new PartitionThroughputInfo(DEFAULT_KB_RATE, DEFAULT_MESSAGE_RATE, "");
```

Comment:
Not sure if it's a good idea to have a default "weight" for partitions that don't have throughput information. And what should be the default value for byte rate and message rate?